File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverFromPreorder(self, traversal: str) -> Optional[TreeNode]:
        st=[]
        i,n=0,len(traversal)
        while i<n:
            d=0
            while i<n and traversal[i]=='-':
                d+=1
                i+=1
            s=i
            while i<n and traversal[i].isdigit():
                i+=1
            num=int(traversal[s:i])
            nd=TreeNode(num)
            while len(st)>d:
                st.pop()
            if st:
                if not st[-1].left:
                    st[-1].left=nd
                else:
                    st[-1].right=nd
            st.append(nd)
        return st[0]

File: 1093-recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode recoverFromPreorder(String traversal) {
        Stack<TreeNode> st=new Stack<>();
        int i=0,n=traversal.length();
        while(i<n)
        {
            int d=0;
            while(i<n && traversal.charAt(i)=='-')
            {
                d+=1;
                i+=1;
            }
            String s="";
            while(i<n && Character.isDigit(traversal.charAt(i)))
            {
                s+=traversal.charAt(i);
                i+=1;
            }
            int num=Integer.parseInt(s);
            TreeNode nd=new TreeNode(num);
            while(st.size()>d)
            {
                st.pop();
            }
            if(!st.isEmpty())
            {
                if(st.peek().left==null)
                st.peek().left=nd;
                else
                st.peek().right=nd;
            }
            st.push(nd);
        }
        return st.get(0);
    }
}